test/storage/cache_response.cpp

```diff
@@ -50,3 +50,175 @@ TEST_F(Storage, CacheResponse) {
 
     loop.run();
 }
+
+// Make sure we don't store a bad server response into the cache
+TEST_F(Storage, CacheNotFound) {
+    SCOPED_TEST(CacheNotFound);
+
+    using namespace mbgl;
+
+    util::RunLoop loop;
+    SQLiteCache cache(":memory:");
+    OnlineFileSource fs(&cache);
+
+    const Resource resource{ Resource::Unknown, "http://127.0.0.1:3000/not-found" };
+
+    // Insert existing data into the cache that will be marked as stale.
+    auto response = std::make_shared<Response>();
+    response->data = std::make_shared<const std::string>("existing data");
+    cache.put(resource, response, FileCache::Hint::Full);
+
+    std::unique_ptr<FileRequest> req1;
+    std::unique_ptr<WorkRequest> req2;
+
+    int counter = 0;
+
+    // Then, request the actual URL and check that we're getting the rigged cache response first,
+    // then the connection error message.
+    req1 = fs.request(resource, [&](Response res) {
+        if (counter == 0) {
+            EXPECT_EQ(nullptr, res.error);
+            EXPECT_EQ(true, res.stale);
+            ASSERT_TRUE(res.data.get());
+            EXPECT_EQ("existing data", *res.data);
+            EXPECT_EQ(Seconds::zero(), res.expires);
+            EXPECT_EQ(Seconds::zero(), res.modified);
+            EXPECT_EQ("", res.etag);
+        } else if (counter == 1) {
+            EXPECT_NE(nullptr, res.error);
+            EXPECT_EQ(Response::Error::Reason::NotFound, res.error->reason);
+            ASSERT_TRUE(res.data.get());
+            EXPECT_EQ("Not Found!", *res.data);
+            req1.reset();
+
+            // Finally, check the cache to make sure we cached the 404 response.
+            req2 = cache.get(resource, [&](std::unique_ptr<Response> res2) {
+                EXPECT_NE(nullptr, res2->error);
+                EXPECT_EQ(Response::Error::Reason::NotFound, res2->error->reason);
+                ASSERT_TRUE(res2->data.get());
+                EXPECT_EQ("Not Found!", *res2->data);
+                CacheNotFound.finish();
+                loop.stop();
+            });
+        } else {
+            FAIL() << "Got too many responses";
+        }
+        counter++;
+    });
+
+    loop.run();
+}
+
+// Make sure we don't store a bad server response into the cache
```

Comment:
"Make sure we don't cache an error response when we can't connect to the server"